mysql-test/suite/innodb/t/innodb-alter-debug.test

```diff
@@ -77,3 +77,70 @@ drop table t1;
 
 # Wait till all disconnects are completed
 --source include/wait_until_count_sessions.inc
+
+--echo #
+--echo # MDEV-11415 AVOID INTERMEDIATE COMMIT WHILE DOING
+--echo #            ALTER TABLE...ALGORITHM=COPY
+--echo #
+
+CREATE DATABASE alter_db;
+USE alter_db;
+
+CREATE TABLE t(a SERIAL, b INT, c INT, d INT) ENGINE=InnoDB;
+INSERT INTO t(a) VALUES(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0);
+INSERT INTO t(a) SELECT a+16 FROM t;
+INSERT INTO t(a) SELECT a+33 FROM t;
+INSERT INTO t(a) VALUES(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0);
+INSERT INTO t(a) SELECT a+81 FROM t;
+INSERT INTO t(a) SELECT a+162 FROM t;
+INSERT INTO t(a) SELECT a+324 FROM t;
+INSERT INTO t(a) SELECT a+648 FROM t;
+INSERT INTO t(a) SELECT a+1296 FROM t;
+INSERT INTO t(a) SELECT a+2592 FROM t;
+INSERT INTO t(a) SELECT a+5186 FROM t;
+INSERT INTO t(a) VALUES(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0),(0);
+
+select count(*) from t;
+
+# try to make the to-be-created secondary index keys randomly distributed
+UPDATE t SET b=a%7, c=a%11, d=a%13;
+# crash right after the last write_row(), before the first commit of ALTER TABLE
+--source include/expect_crash.inc
+
+SET DEBUG_DBUG='+d,crash_copy_before_commit';
+--error 2013
+# create 32 secondary indexes
+ALTER TABLE t ADD INDEX(b,c,d,a),ADD INDEX(b,c,a,d),ADD INDEX(b,a,c,d),ADD INDEX(b,a,d,c),
+              ADD INDEX(b,d,a,c),ADD INDEX(b,d,c,a),ADD INDEX(a,b,c,d),ADD INDEX(a,b,d,c),
+              ADD INDEX(a,c,b,d),ADD INDEX(a,c,d,b),ADD INDEX(a,d,b,c),ADD INDEX(a,d,c,b),
+              ADD INDEX(c,a,b,d),ADD INDEX(c,a,d,b),ADD INDEX(c,b,a,d),ADD INDEX(c,b,d,a),
+              ADD INDEX(c,d,a,b),ADD INDEX(c,d,b,a),ADD INDEX(d,a,b,c),ADD INDEX(d,a,c,b),
+              ADD INDEX(d,b,a,c),ADD INDEX(d,b,c,a),ADD INDEX(d,c,a,b),ADD INDEX(d,c,b,a),
+              ADD INDEX(a,b,c), ADD INDEX(a,c,b), ADD INDEX(a,c,d), ADD INDEX(a,d,c),
+              ADD INDEX(a,b,d), ADD INDEX(a,d,b), ADD INDEX(b,c,d), ADD INDEX(b,d,c),
+              ALGORITHM=COPY;
+
+--source include/start_mysqld.inc
```

Comment:
Actually, now I think that I know what the problem is. I had no problem with the test yesterday, because the code changes were present. Without the code changes, the test is revealing a recent regression that I introduced. The reason is that there would be undo log records for inserting the individual rows, and therefore the following sequence of actions would take place:
```
2018-01-17 14:04:13 140541470619584 [Note] InnoDB: 1 transaction(s) which must be rolled back or cleaned up in total 788 row operations to undo
2018-01-17 14:04:13 140541470619584 [Note] InnoDB: Trx id counter is 2304
2018-01-17 14:04:13 140541470619584 [Note] InnoDB: Starting final batch to recover 113 pages from redo log.
2018-01-17 14:04:14 140541470619584 [Note] InnoDB: Deferring DROP TABLE `test`.`#sql-43f8_8`; renaming to test/#sql-ib20-4145398497
2018-01-17 14:04:14 0x7fd25c6eb7c0  InnoDB: Assertion failure in file /mariadb/10.2-ext/storage/innobase/dict/dict0dict.cc line 5397
InnoDB: Failing assertion: trx->mysql_thd
```
With your code changes, no locks would be recovered for the `#sql-43f8_8` table that was created for the `ALTER TABLE` operation, and therefore we would be able to drop the ‘garbage’ table immediately, without any waiting.

Reply:
I pushed the fix in commit 656f66def27b7a2cf42a28f873f1eeef0416aa71